File: tests/unit/math/test_quaternion.py
import math
import sys
from pathlib import Path

import numpy as np
import pytest

ROOT = Path(__file__).resolve().parents[3]
sys.path.insert(0, str(ROOT / "build"))

try:
    import saltro_py
except ImportError as e:
    pytest.skip(f"saltro Python bindings not available: {e}", allow_module_level=True)


def fixed_unit_quat():
    q = np.array([0.8, 0.2, -0.4, 0.3], dtype=float)
    return q / np.linalg.norm(q)


def fixed_vec():
    return np.array([1.3, -0.7, 2.1], dtype=float)


def quat_mul(p, r):
    return np.array([
        p[0] * r[0] - p[1] * r[1] - p[2] * r[2] - p[3] * r[3],
        p[0] * r[1] + p[1] * r[0] + p[2] * r[3] - p[3] * r[2],
        p[0] * r[2] - p[1] * r[3] + p[2] * r[0] + p[3] * r[1],
        p[0] * r[3] + p[1] * r[2] - p[2] * r[1] + p[3] * r[0],
    ])


def perturb_on_sphere(q, theta):
    n = np.linalg.norm(theta)
    if n < 1e-300:
        dq = np.array([1.0, 0.0, 0.0, 0.0])
    else:
        dq = np.zeros(4)
        dq[0] = np.cos(n / 2.0)
        dq[1:] = np.sin(n / 2.0) * theta / n
    return quat_mul(q, dq)


def test_normalize_quat_produces_unit_norm():
    q = np.array([2.0, -1.0, 0.5, 3.0])
    qn = saltro_py.normalizeQuat(q)
    assert abs(np.linalg.norm(qn) - 1.0) < 1e-12
    assert abs(np.linalg.norm(qn * np.linalg.norm(q) - q)) < 1e-12


def test_normalize_quat_preserves_unit_quaternion():
    q = fixed_unit_quat()
    qn = saltro_py.normalizeQuat(q)
    assert np.linalg.norm(qn - q) < 1e-12


def test_normalize_quat_throws_on_near_zero_quaternion():
    with pytest.raises(Exception):
        saltro_py.normalizeQuat(np.array([1e-15, 0.0, -1e-15, 0.0]))


def test_rotation_matrix_is_orthonormal_with_det_plus_one():
    r = saltro_py.rotationMatrix(fixed_unit_quat())
    assert np.linalg.norm(r @ r.T - np.eye(3)) < 1e-12
    assert abs(np.linalg.det(r) - 1.0) < 1e-12


def test_rotation_matrix_normalizes_input():
    q = fixed_unit_quat()
    r1 = saltro_py.rotationMatrix(q)
    r2 = saltro_py.rotationMatrix(2.5 * q)
    assert np.linalg.norm(r1 - r2) < 1e-12


def test_rotation_matrix_identity_quaternion_gives_identity():
    r = saltro_py.rotationMatrix(np.array([1.0, 0.0, 0.0, 0.0]))
    assert np.linalg.norm(r - np.eye(3)) < 1e-12


def test_rotation_matrix_known_90deg_about_z():
    c = np.cos(np.pi / 4.0)
    s = np.sin(np.pi / 4.0)
    r = saltro_py.rotationMatrix(np.array([c, 0.0, 0.0, s]))
    expected = np.array([[0.0, -1.0, 0.0],
                         [1.0, 0.0, 0.0],
                         [0.0, 0.0, 1.0]])
    assert np.linalg.norm(r - expected) < 1e-12


def test_skew_symmetric_implements_cross_product():
    a = np.array([1.0, -2.0, 3.0])
    b = np.array([-0.5, 0.4, 1.7])
    assert np.linalg.norm(np.cross(a, b) - saltro_py.skewSymmetric(a) @ b) < 1e-14


def test_skew_symmetric_is_antisymmetric_with_zero_diagonal():
    s = saltro_py.skewSymmetric(np.array([1.3, -0.7, 2.1]))
    assert np.linalg.norm(s + s.T) < 1e-14
    assert abs(s[0, 0]) < 1e-14
    assert abs(s[1, 1]) < 1e-14
    assert abs(s[2, 2]) < 1e-14


def test_quat_norm_jacobian_matches_finite_difference_first_three_columns():
    q = np.array([0.8, 0.2, -0.4, 0.3], dtype=float)
    janalytic = saltro_py.quatNormJacobian(q)
    eps = 1e-6
    jnum = np.zeros((4, 4))
    for j in range(4):
        qp = q.copy()
        qm = q.copy()
        qp[j] += eps
        qm[j] -= eps
        fp = qp / np.linalg.norm(qp)
        fm = qm / np.linalg.norm(qm)
        jnum[:, j] = (fp - fm) / (2.0 * eps)
    assert np.max(np.abs(janalytic - jnum[:, :3])) < 1e-6


def test_find_w_mat_satisfies_quaternion_kinematics():
    q = fixed_unit_quat()
    omega = np.array([0.13, -0.07, 0.21])
    w = saltro_py.findWMat(q)

    q0 = q[0]
    qv = q[1:]
    qdot_ref = np.zeros(4)
    qdot_ref[0] = -0.5 * np.dot(qv, omega)
    qdot_ref[1:] = 0.5 * (q0 * omega + np.cross(qv, omega))

    qdot = 0.5 * (w @ omega)
    assert np.linalg.norm(qdot - qdot_ref) < 1e-12


def test_find_w_mat_columns_are_orthogonal_to_q():
    q = fixed_unit_quat()
    w = saltro_py.findWMat(q)
    assert np.linalg.norm(w.T @ q) < 1e-12


def test_drotmat_t_vec_dq_matches_on_manifold_finite_differences():
    qs = [np.array([1.0, 0.0, 0.0, 0.0]), fixed_unit_quat()]
    v = fixed_vec()

    for q in qs:
        j = saltro_py.drotmatTvecdq(q, v)
        w = saltro_py.findWMat(q)
        analytic = j.T @ (0.5 * w)

        eps = 1e-7
        numerical = np.zeros((3, 3))
        for k in range(3):
            tp = np.zeros(3)
            tm = np.zeros(3)
            tp[k] = eps
            tm[k] = -eps
            fp = saltro_py.rotationMatrix(perturb_on_sphere(q, tp)).T @ v
            fm = saltro_py.rotationMatrix(perturb_on_sphere(q, tm)).T @ v
            numerical[:, k] = (fp - fm) / (2.0 * eps)

        assert np.max(np.abs(analytic - numerical)) < 1e-6


def test_ddrotmat_t_vec_dqdq_slices_are_symmetric():
    h = saltro_py.ddrotmatTvecdqdq(fixed_unit_quat(), fixed_vec())
    for block in h:
        assert np.linalg.norm(block - block.T) < 1e-12


def test_ddrotmat_t_vec_dqdq_matches_finite_differences_of_drotmat():
    q = fixed_unit_quat()
    v = fixed_vec()
    h_analytic = saltro_py.ddrotmatTvecdqdq(q, v)

    eps = 1e-6
    h_num = [np.zeros((4, 4)), np.zeros((4, 4)), np.zeros((4, 4))]
    for b in range(4):
        qp = q.copy()
        qm = q.copy()
        qp[b] += eps
        qm[b] -= eps
        jp = saltro_py.drotmatTvecdq(qp, v)
        jm = saltro_py.drotmatTvecdq(qm, v)
        dj = (jp - jm) / (2.0 * eps)
        for k in range(3):
            h_num[k][:, b] = dj[:, k]

    for k in range(3):
        assert np.max(np.abs(h_analytic[k] - h_num[k])) < 1e-5


def test_ddrotmat_t_vec_dqdq_finite_difference_reference_is_symmetric():
    q = fixed_unit_quat()
    v = fixed_vec()
    eps = 1e-6
    h_num = [np.zeros((4, 4)), np.zeros((4, 4)), np.zeros((4, 4))]
    for b in range(4):
        qp = q.copy()
        qm = q.copy()
        qp[b] += eps
        qm[b] -= eps
        dj = (saltro_py.drotmatTvecdq(qp, v) - saltro_py.drotmatTvecdq(qm, v)) / (2.0 * eps)
        for k in range(3):
            h_num[k][:, b] = dj[:, k]

    for k in range(3):
        assert np.linalg.norm(h_num[k] - h_num[k].T) < 1e-5

File: tests/unit/math/test_quaternion.cpp
#include <catch2/catch_test_macros.hpp>
#include <Eigen/Dense>
#include <array>
#include <cmath>

#include <saltro/math/quaternion.h>

using namespace saltro::math;

namespace {

// A fixed, non-trivial unit quaternion used across tests.
Vec4 fixedUnitQuat() {
    Vec4 q(0.8, 0.2, -0.4, 0.3);
    return q / q.norm();
}

// A fixed, non-trivial vector to be rotated.
Vec3 fixedVec() {
    return Vec3(1.3, -0.7, 2.1);
}

// Scalar-first Hamilton product p (x) r.
Vec4 quatMul(const Vec4& p, const Vec4& r) {
    Vec4 out;
    out(0) = p(0) * r(0) - p(1) * r(1) - p(2) * r(2) - p(3) * r(3);
    out(1) = p(0) * r(1) + p(1) * r(0) + p(2) * r(3) - p(3) * r(2);
    out(2) = p(0) * r(2) - p(1) * r(3) + p(2) * r(0) + p(3) * r(1);
    out(3) = p(0) * r(3) + p(1) * r(2) - p(2) * r(1) + p(3) * r(0);
    return out;
}

// Perturb a unit quaternion by a body-frame rotation vector (stays on the
// unit sphere): q' = q (x) exp(theta/2).
Vec4 perturbOnSphere(const Vec4& q, const Vec3& theta) {
    double n = theta.norm();
    Vec4 dq;
    if (n < 1e-300) {
        dq << 1.0, 0.0, 0.0, 0.0;
    } else {
        dq(0) = std::cos(n / 2.0);
        dq.tail<3>() = std::sin(n / 2.0) * theta / n;
    }
    return quatMul(q, dq);
}

}  // namespace

// ============================================================================
// normalizeQuat
// ============================================================================

TEST_CASE("normalizeQuat produces unit norm", "[math][quaternion][normalize]") {
    Vec4 q(2.0, -1.0, 0.5, 3.0);
    Vec4 qn = normalizeQuat(q);
    REQUIRE(std::abs(qn.norm() - 1.0) < 1e-12);
    // Direction preserved: qn parallel to q.
    REQUIRE(std::abs((qn * q.norm() - q).norm()) < 1e-12);
}

TEST_CASE("normalizeQuat of an already-unit quaternion is unchanged", "[math][quaternion][normalize]") {
    Vec4 q = fixedUnitQuat();
    Vec4 qn = normalizeQuat(q);
    REQUIRE(std::abs((qn - q).norm()) < 1e-12);
}

TEST_CASE("normalizeQuat throws on near-zero quaternion", "[math][quaternion][normalize]") {
    Vec4 q(1e-15, 0.0, -1e-15, 0.0);
    REQUIRE_THROWS(normalizeQuat(q));
}

// ============================================================================
// rotationMatrix
// ============================================================================

TEST_CASE("rotationMatrix is orthonormal with det +1", "[math][quaternion][rotation]") {
    Vec4 q = fixedUnitQuat();
    Mat33 R = rotationMatrix(q);

    Mat33 should_be_I = R * R.transpose();
    REQUIRE((should_be_I - Mat33::Identity()).norm() < 1e-12);
    REQUIRE(std::abs(R.determinant() - 1.0) < 1e-12);
}

TEST_CASE("rotationMatrix normalizes its input", "[math][quaternion][rotation]") {
    Vec4 q = fixedUnitQuat();
    Mat33 R1 = rotationMatrix(q);
    Mat33 R2 = rotationMatrix(2.5 * q);  // same direction, different norm
    REQUIRE((R1 - R2).norm() < 1e-12);
}

TEST_CASE("rotationMatrix identity quaternion gives identity", "[math][quaternion][rotation]") {
    Vec4 q(1.0, 0.0, 0.0, 0.0);
    Mat33 R = rotationMatrix(q);
    REQUIRE((R - Mat33::Identity()).norm() < 1e-12);
}

TEST_CASE("rotationMatrix known 90deg about z", "[math][quaternion][rotation]") {
    // q = [cos(45), 0, 0, sin(45)] -> 90 deg about +z.
    double c = std::cos(M_PI / 4.0);
    double s = std::sin(M_PI / 4.0);
    Vec4 q(c, 0.0, 0.0, s);
    Mat33 R = rotationMatrix(q);

    // With this convention R(0,1) = 2(q1q2 - q0q3) = -2 c s = -sin(90) = -1.
    // R rotates a body vector to the reference frame: R * e_x should be e_y
    // if this is an active rotation by +90 about z. Check the matrix entries.
    Mat33 expected;
    expected << 0.0, -1.0, 0.0,
                1.0,  0.0, 0.0,
                0.0,  0.0, 1.0;
    REQUIRE((R - expected).norm() < 1e-12);
}

// ============================================================================
// skewSymmetric
// ============================================================================

TEST_CASE("skewSymmetric implements cross product", "[math][quaternion][skew]") {
    Vec3 a(1.0, -2.0, 3.0);
    Vec3 b(-0.5, 0.4, 1.7);
    Vec3 cross_direct = a.cross(b);
    Vec3 cross_skew = skewSymmetric(a) * b;
    REQUIRE((cross_direct - cross_skew).norm() < 1e-14);
}

TEST_CASE("skewSymmetric is antisymmetric with zero diagonal", "[math][quaternion][skew]") {
    Vec3 a(1.3, -0.7, 2.1);
    Mat33 S = skewSymmetric(a);
    REQUIRE((S + S.transpose()).norm() < 1e-14);
    REQUIRE(std::abs(S(0, 0)) < 1e-14);
    REQUIRE(std::abs(S(1, 1)) < 1e-14);
    REQUIRE(std::abs(S(2, 2)) < 1e-14);
}

// ============================================================================
// findWMat / quatNormJacobian (FD vs analytic)
// ============================================================================

TEST_CASE("quatNormJacobian matches finite differences of normalizeQuat",
          "[math][quaternion][jacobian][finite-diff]") {
    // quatNormJacobian returns d(q/|q|)/dq. Header type is Mat43 but the
    // implementation fills a 4x3 block; we compare the available 4x3 entries
    // against the (4x4) numerical normalization Jacobian's first 3 columns.
    Vec4 q(0.8, 0.2, -0.4, 0.3);  // raw (unnormalized direction)
    Mat43 Janalytic = quatNormJacobian(q);

    const double eps = 1e-6;
    // Numerical d(normalize)/dq, 4x4.
    Mat44 Jnum = Mat44::Zero();
    for (int j = 0; j < 4; ++j) {
        Vec4 qp = q, qm = q;
        qp(j) += eps;
        qm(j) -= eps;
        Vec4 fp = qp / qp.norm();
        Vec4 fm = qm / qm.norm();
        Jnum.col(j) = (fp - fm) / (2.0 * eps);
    }

    // Compare the 4x3 analytic block against the first 3 numerical columns.
    for (int i = 0; i < 4; ++i) {
        for (int j = 0; j < 3; ++j) {
            REQUIRE(std::abs(Janalytic(i, j) - Jnum(i, j)) < 1e-6);
        }
    }
}

TEST_CASE("findWMat satisfies quaternion kinematics qdot = 0.5 W omega",
          "[math][quaternion][wmat]") {
    Vec4 q = fixedUnitQuat();
    Vec3 omega(0.13, -0.07, 0.21);
    Mat43 W = findWMat(q);

    // Reference qdot from the standard scalar-first kinematics:
    // qdot = 0.5 * [ -qv . w ; q0 w + qv x w ].
    double q0 = q(0);
    Vec3 qv = q.tail<3>();
    Vec4 qdot_ref;
    qdot_ref(0) = -0.5 * qv.dot(omega);
    qdot_ref.tail<3>() = 0.5 * (q0 * omega + qv.cross(omega));

    Vec4 qdot = 0.5 * W * omega;
    REQUIRE((qdot - qdot_ref).norm() < 1e-12);
}

TEST_CASE("findWMat columns are orthogonal to q (W^T q = 0)",
          "[math][quaternion][wmat]") {
    Vec4 q = fixedUnitQuat();
    Mat43 W = findWMat(q);
    Vec3 wtq = W.transpose() * q;
    REQUIRE(wtq.norm() < 1e-12);
}

// ============================================================================
// drotmatTvecdq (FD of R(q)^T v vs analytic)
// ============================================================================

// drotmatTvecdq returns a 4x3 Euclidean Jacobian J with J(a, k) =
// d[R(q)^T v]_k / dq_a, but as a *unit-norm-constrained* derivative (it carries
// the q0 normalization contribution, so a naive Euclidean FD of the raw 3-2-1
// rotation-matrix formula does NOT match it). The physically meaningful and
// codebase-consistent check is the ON-MANIFOLD derivative: contracting J onto
// the quaternion tangent space via findWMat must reproduce the finite-difference
// of R(q)^T v under unit-norm-preserving (body-rotation) perturbations.
TEST_CASE("drotmatTvecdq matches on-manifold finite differences of R(q)^T v",
          "[math][quaternion][drotmat][finite-diff]") {
    const std::array<Vec4, 2> qs = {Vec4(1.0, 0.0, 0.0, 0.0), fixedUnitQuat()};
    Vec3 v = fixedVec();

    for (const Vec4& q : qs) {
        Mat43 J = drotmatTvecdq(q, v);
        Mat43 W = findWMat(q);
        // d(R^T v)/dtheta = J^T (dq/dtheta) = J^T (0.5 W), shape 3x3.
        Mat33 analytic = J.transpose() * (0.5 * W);

        const double eps = 1e-7;
        Mat33 numerical = Mat33::Zero();
        for (int k = 0; k < 3; ++k) {
            Vec3 tp = Vec3::Zero(), tm = Vec3::Zero();
            tp(k) = eps;
            tm(k) = -eps;
            Vec3 fp = rotationMatrix(perturbOnSphere(q, tp)).transpose() * v;
            Vec3 fm = rotationMatrix(perturbOnSphere(q, tm)).transpose() * v;
            numerical.col(k) = (fp - fm) / (2.0 * eps);
        }

        double maxerr = (analytic - numerical).cwiseAbs().maxCoeff();
        INFO("q = " << q.transpose());
        INFO("max |analytic - numerical| = " << maxerr);
        INFO("analytic=\n" << analytic);
        INFO("numerical=\n" << numerical);
        REQUIRE(maxerr < 1e-6);
    }
}

// ============================================================================
// ddrotmatTvecdqdq -- the helper that had the bug; pin it hard.
// ============================================================================

// =========================================================================
// KNOWN BUG (flagged, NOT papered over):
//
// ddrotmatTvecdqdq is the Euclidean second derivative of R(q)^T v, indexed in
// satellite.cpp as H[i](QUAT_INDEX+j, QUAT_INDEX+k) = d^2(R^T v)_i / (dq_j dq_k).
// As a Hessian it must be (1) symmetric in (j,k) and (2) equal to the Euclidean
// Jacobian of drotmatTvecdq (the matching first derivative). These cases pin
// both. The qv-qv block of this helper was previously incomplete and asymmetric
// (wrong by O(1)); it is exercised here at a NON-identity quaternion, where the
// defect shows -- identity-only tests cancel it. (This is the direct regression
// test for the rotation-Hessian fix.)
// =========================================================================

TEST_CASE("ddrotmatTvecdqdq slices are symmetric in the two q-indices",
          "[math][quaternion][ddrotmat][symmetry]") {
    Vec4 q = fixedUnitQuat();
    Vec3 v = fixedVec();
    std::array<Mat44, 3> H = ddrotmatTvecdqdq(q, v);
    for (int k = 0; k < 3; ++k) {
        double asym = (H[static_cast<size_t>(k)] - H[static_cast<size_t>(k)].transpose()).norm();
        INFO("component k=" << k << " asymmetry norm = " << asym);
        REQUIRE(asym < 1e-12);
    }
}

TEST_CASE("ddrotmatTvecdqdq matches finite differences of drotmatTvecdq",
          "[math][quaternion][ddrotmat][finite-diff]") {
    // H[k](a,b) = d^2 (R^T v)_k / (dq_a dq_b).
    // drotmatTvecdq returns J with J(a,k) = d(R^T v)_k/dq_a. So
    // d J(a,k) / dq_b = H[k](a,b). We finite-difference drotmatTvecdq and the
    // analytic helper must match it (and the FD reference is itself symmetric to
    // ~1e-10, see the positive-control case below).
    Vec4 q = fixedUnitQuat();
    Vec3 v = fixedVec();
    std::array<Mat44, 3> Hanalytic = ddrotmatTvecdqdq(q, v);

    const double eps = 1e-6;
    std::array<Mat44, 3> Hnum = {Mat44::Zero(), Mat44::Zero(), Mat44::Zero()};
    for (int b = 0; b < 4; ++b) {
        Vec4 qp = q, qm = q;
        qp(b) += eps;
        qm(b) -= eps;
        Mat43 Jp = drotmatTvecdq(qp, v);
        Mat43 Jm = drotmatTvecdq(qm, v);
        Mat43 dJ = (Jp - Jm) / (2.0 * eps);  // dJ(a,k) = dJ(a,k)/dq_b
        for (int k = 0; k < 3; ++k) {
            for (int a = 0; a < 4; ++a) {
                Hnum[static_cast<size_t>(k)](a, b) = dJ(a, k);
            }
        }
    }

    for (int k = 0; k < 3; ++k) {
        double maxerr = (Hanalytic[static_cast<size_t>(k)] - Hnum[static_cast<size_t>(k)])
                            .cwiseAbs()
                            .maxCoeff();
        INFO("component k=" << k << " max |analytic - numerical| = " << maxerr);
        INFO("analytic=\n" << Hanalytic[static_cast<size_t>(k)]);
        INFO("numerical=\n" << Hnum[static_cast<size_t>(k)]);
        REQUIRE(maxerr < 1e-5);
    }
}

// Positive control: the numerical Hessian (FD of drotmatTvecdq) is itself
// symmetric, proving the symmetry failure above is a defect in the analytic
// helper and not a property of the underlying function.
TEST_CASE("ddrotmatTvecdqdq finite-difference reference is symmetric",
          "[math][quaternion][ddrotmat][symmetry]") {
    Vec4 q = fixedUnitQuat();
    Vec3 v = fixedVec();
    const double eps = 1e-6;
    std::array<Mat44, 3> Hnum = {Mat44::Zero(), Mat44::Zero(), Mat44::Zero()};
    for (int b = 0; b < 4; ++b) {
        Vec4 qp = q, qm = q;
        qp(b) += eps;
        qm(b) -= eps;
        Mat43 dJ = (drotmatTvecdq(qp, v) - drotmatTvecdq(qm, v)) / (2.0 * eps);
        for (int k = 0; k < 3; ++k)
            for (int a = 0; a < 4; ++a)
                Hnum[static_cast<size_t>(k)](a, b) = dJ(a, k);
    }
    for (int k = 0; k < 3; ++k) {
        double asym = (Hnum[static_cast<size_t>(k)] - Hnum[static_cast<size_t>(k)].transpose()).norm();
        INFO("component k=" << k << " FD asymmetry norm = " << asym);
        REQUIRE(asym < 1e-5);
    }
}
